pkg/parser/uri.go

```diff
@@ -0,0 +1,46 @@
+package parser
+
+import (
+	"errors"
+	"net/url"
+	"strings"
+)
+
+var (
+	// ErrEmptyDSN defines error returned when no dsn is provided
+	ErrEmptyDSN = errors.New("Empty string provided for dsn")
+)
+
+// If our dsn scheme is not immediately followed by a forward slash (/),
+// then we use Opaque as our path
+//
+func parseURI(dsn string) (*url.URL, error) {
+	uri, err := url.Parse(dsn)
+	if err != nil {
+		return nil, err
+	}
+	if dsn == "" {
+		return nil, ErrEmptyDSN
+	}
+
+	schemeHasForwardSlash := checkSchemeSuffix(dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	if !schemeHasForwardSlash {
+		uri.Path = uri.Opaque
+	}
+
+	return uri, err
+}
+
+// Checks if a dsn scheme contains a forward slash immediately
+// after the first colon
+func checkSchemeSuffix(dsn string) bool {
```

Comment:
Why do we need this function at all? If you don't have a proper `/` the parse function returns an error for you that states that an invalid URL was provided. Is there any case that I'm not seeing here?

Reply:
No, not always. In at least two cases, the parser does not know that we want to use the element in the `url.Host` position as our path, so doesn't save  that element into `url.Path`. Rather, it saves it to `url.Opaque`.  It is not an error and so will not return any error. We need a way to know if we're dealing with this specific kind of dsn.